src/storage/src/hummock/file_cache/store.rs

```diff
@@ -12,431 +12,533 @@
+    K: Key,
+    V: Value,
 {
-    fn new(
-        store: &'a Store<K, V>,
-        block_size: usize,
-        buffer_capacity: usize,
-        item_capacity: usize,
-        max_write_size: usize,
-    ) -> Self {
-        Self {
-            keys: Vec::with_capacity(item_capacity),
-            buffers: Vec::with_capacity(item_capacity),
-            blocs: Vec::with_capacity(item_capacity),
-            data_len: 0,
-
-            block_size,
-            buffer_capacity,
-            max_write_size,
-
-            store,
-        }
-    }
-
-    #[allow(clippy::uninit_vec)]
-    pub fn append<'b>(&'b mut self, key: K, value: &V) {
-        let offset = self.data_len;
-        let len = value.encoded_len();
-        let bloc = BlockLoc {
-            bidx: offset as u32 / self.block_size as u32,
-            len: len as u32,
-        };
-        self.blocs.push(bloc);
-
-        let rotate_last_mut = |buffers: &'b mut Vec<_>| {
-            buffers.push(DioBuffer::with_capacity_in(
-                self.buffer_capacity,
-                &DIO_BUFFER_ALLOCATOR,
-            ));
-            buffers.last_mut().unwrap()
-        };
+    pub dir: PathBuf,
+    pub capacity: usize,
+    pub file_capacity: usize,
+    pub buffer_pool_size: usize,
+    pub device_align: usize,
+    pub device_io_size: usize,
+    pub flushers: usize,
+    pub flush_rate_limit: usize,
+    pub reclaimers: usize,
+    pub reclaim_rate_limit: usize,
+    pub recover_concurrency: usize,
+    pub lfu_window_to_cache_size_ratio: usize,
+    pub lfu_tiny_lru_capacity_ratio: f64,
+    pub rated_random_rate: usize,
+    pub prometheus_registry: Option<Registry>,
+    pub prometheus_namespace: Option<String>,
+    pub event_listener: Vec<Arc<dyn EventListener<K = K, V = V>>>,
+    pub enable_filter: bool,
+}
 
-        let buffer = match self.buffers.last_mut() {
-            Some(buffer) if buffer.len() + len > self.max_write_size => {
-                rotate_last_mut(&mut self.buffers)
-            }
-            None => rotate_last_mut(&mut self.buffers),
-            Some(buffer) => buffer,
-        };
+pub struct FoyerRuntimeConfig<K, V>
+where
+    K: Key,
+    V: Value,
+{
+    pub foyer_store_config: FoyerStoreConfig<K, V>,
+    pub runtime_worker_threads: Option<usize>,
+}
 
-        let buffer_offset = buffer.len();
-        let buffer_len = utils::align_up(self.block_size, len);
-        buffer.reserve(buffer_offset + buffer_len);
-        unsafe {
-            buffer.set_len(buffer_offset + buffer_len);
-        }
-        value.encode(&mut buffer[buffer_offset..buffer_offset + buffer_len]);
-        self.data_len += buffer_len;
+#[derive(Debug, Clone, Copy, PartialEq, PartialOrd, Eq, Ord, Hash)]
+pub struct SstableBlockIndex {
+    pub sst_id: HummockSstableObjectId,
+    pub block_idx: u64,
+}
 
-        self.keys.push(key);
+impl Key for SstableBlockIndex {
+    fn serialized_len(&self) -> usize {
+        8 + 8 // sst_id (8B) + block_idx (8B)
```

Comment:
`sizeof::<HummockSstableObjectId>()`?

Reply:
I prefer not. `size_of` calculates its memory size, but the serialized size is needed here. 🤔